src/LSST/state/StateMap.cpp

```diff
@@ -0,0 +1,203 @@
+    _safeStates.push_back(_standbyState);
+}
+
+void StateMap::insertIntoMap(State::Ptr const& state) {
+    auto [iter, success] = _stateMap.emplace(state->getName(), state);
+    if (!success) {
+        string name = state->getName();
+        throw util::Bug(ERR_LOC, name + " was already in stateMap!");
+    }
+}
+
+bool StateMap::changeState(string const& newState) {
+    LDEBUG("changeState newState=", newState);
+
+    // Find the state
+    auto iter = _stateMap.find(newState);
+    if (iter == _stateMap.end()) {
+        LERROR("changeState unknown newState=", newState);
+        return false;
+    }
+    auto& newStatePtr = iter->second;
+    return _changeState(newStatePtr);
+}
+
+bool StateMap::changeState(State::Ptr const& newState) {
+    if (newState == nullptr) {
+        LWARN("StateMap cannot change state to nullptr");
+        return false;
+    }
+
+    // The new state much match a state in the state map.
+    // Find the state
+    auto iter = _stateMap.find(newState->getName());
+    if (iter == _stateMap.end()) {
+        LERROR("changeState unknown newState=", newState);
+        return false;
+    }
+    if (iter->second != newState) {
+        LWARN("StateMap newState=", newState->getName(), " is not in the map");
+        return false;
+    }
+    return _changeState(newState);
+}
+
+bool StateMap::_changeState(State::Ptr const& newState) {
+    if (newState == nullptr) {
+        LWARN("StateMap cannot change state to nullptr");
+        return false;
+    }
+
+    // The new state should have been verified as being in the map.
+    auto oldState = _currentState;
+    if (newState == oldState) {
+        // Re-run the enter state options. This is useful for for
+        // states like StandbyState to make sur everything is turned off.
+        // Other states should be checking the old state and taking appropriate
+        // action.
+        newState->onEnterState(oldState);
+        return true;
+    }
+
+    // If the current state is "startupState", the state cannot be changed until
+    // the Model believes all files have been loaded, etc.
+    if (_currentState->getName() == "StartupState") {
+        if (!_startupState->isStartupFinished()) {
+            LERROR("StateMap::_changeState cannot leave startupState as system isn't ready.");
+            return false;
+        }
+    }
+
+    // Safe state transitions come from "Support_VIs->States" in the LabView code.
+    // Each state allows only certain actions as seen in the "Support_VIs->Commands"
+    // section where several Commands exec.vi actions depend on the _currentState,
+    // such as "Support_VIs->Commands->Pause.lvclass:exec.vi"
+
+    bool acceptable = false;
+    auto const currentStateName = _currentState->getName();
+
+    if (isASafeState(newState->getName())) {
+        // It's always safe to go to these states as they
+        // turn off power and motion.
+        acceptable = true;
+    } else if (currentStateName == "IdleState") {

```

Comment:
Why not use the `switch` statement instead? I thought it should be more clear compared with `if` statement in this part?

Reply:
Can't do a switch statement on strings. I intend to change that to use enums, and then a switch statement can be used. Using the strings was an experiment that has not really worked out.

Reply:
Yep~ I thought this as the `match` in Python. Sorry about that. Thanks!